packages/core/src/index.ts

```diff
@@ -1,89 +1,185 @@
-      }
-    }
-  };
-
   const runHook = (hook: string, args: any) => {
     try {
       return getHooks(hook, res, args);
     } catch (e) {
       sendError(`[EXCEPTION IN REGISTERED HOOK ${hook}]`, e);
     }
   };
-
-  const sendError = (subject: string, e: Error = null) => {
+  /**
+   * Centralized api error handling
+   * @param {subject} string subject 
+   * @param {e} NodeJS.ErrnoException error
+   */
+  const sendError = (subject: string, e: NodeJS.ErrnoException = null) => {
     const exception = e ? e.stack || e.message : "";
     const message = subject + "\n\n\n" + exception;
-    if (!res.headersSent) {
-      res.status(500).json({
-        zenroom_errors: zenroom_errors,
-        result: zenroom_result,
-        exception: message,
+    if (e.code === "ENOENT") {
+      getMessage(req).then( (mes)=>{
+        res.status(404).send(mes);
       });
-      if (e) next(e);
+    } else{
+    if (!res.headersSent) {
+        res.status(500).json({
+          zenroom_errors: zenroom_errors,
+          result: zenroom_result,
+          exception: message,
+        });
+        if (e) next(e);
     }
+   }
   };
 
+  /**
+   * Centralized api response handling
+   * @param {restroomResult} RestroomResult containing restroom result 
+   * @param {res} Response endpoint response
+   */
+  const buildEndpointResponse = (restroomResult: RestroomResult, res: Response) => {
+    if (restroomResult?.error) {
+      sendError(restroomResult.errorMessage, restroomResult.error);
+    } else {
+      res.status(restroomResult.status).json(restroomResult?.result);
+    }
+  };
+
+  async function resolveRestroomResult(restroomResult: RestroomResult): Promise<RestroomResult> {
+    return new Promise((resolve) => {
+      resolve(restroomResult);
+    });
+  };
+
+  /**
+   * Function responsible to execute the chain
+   * @param {ymlFile} string containing restroom result 
+   * @param {data} object data object coming from endpoint 
+   * @param {keys} object keys object coming from file 
+   */
+  async function executeChain(fileContents: string, data: any): Promise<any> {
+    const ymlContent: any = yaml.load(fileContents);
+    const startBlock: string = ymlContent.start;
+  
+    return await evaluateBlock(startBlock, ymlContent, data);
+  }
+
+  const getRestroomResult = async (contractName:string, data:any) : Promise<RestroomResult> => {
+    const isChain = contractName.split(".")[1] === 'chain' || false;
+    const keys = isChain ? "{}" : getKeys(contractName);
+    try {
+      return isChain ? executeChain(getYml(contractName.split(".")[0]), data) : callRestroom(data, keys, getConf(contractName), getContractByContractName(contractName), contractName);
+    } catch (err){
+      return await resolveRestroomResult({
+        error: err
+      });
+    }
+  }
+
+  async function evaluateBlock(
+    block: string,
+    ymlContent: any,
+    data: any
+  ): Promise<RestroomResult> {
+    console.log("Current block is " + block);
```

Comment:
Maybe need some debug thing, like https://www.npmjs.com/package/debug to be able to suppress this console.logs in case... easy to forget them and in embedded devices logging consumes space fastly

Reply:
cheddite?